common/store.go

```diff
@@ -0,0 +1,265 @@
+	Connect() error
+	// Create the schema, dropping existing one if the true flag is specified
+	CreateSchema(force bool) error
+}
+
+// ServiceStore interface is what each service's store needs to implement.
+type ServiceStore interface {
+	// Entities returns list of entities (DB tables) this store is managing.
+	Entities() []interface{}
+	// CreateSchemaPostProcess runs whatever required post-processing after
+	// schema creation (perhaps initializing DB with some initial or sample data).
+	CreateSchemaPostProcess() error
+}
+
+// Function implementing createSchema functionality
+type createSchema func(dbStore DbStore, force bool) error
+
+type DbStore struct {
+	ServiceStore      ServiceStore
+	Config            *StoreConfig
+	Db                *gorm.DB
+	createSchemaFuncs map[string]createSchema
+}
+
+// SetConfig sets the config object from a map.
+func (dbStore *DbStore) SetConfig(configMap map[string]interface{}) error {
+	config := makeStoreConfig(configMap)
+	dbStore.Config = &config
+	dbStore.createSchemaFuncs = make(map[string]createSchema)
+	dbStore.createSchemaFuncs["mysql"] = createSchemaMysql
+	dbStore.createSchemaFuncs["sqlite3"] = createSchemaSqlite3
+	return nil
+}
+
+// GetPasswordFunction returns appropriate function to hash
+// password depending on the underlying DB (note that in sqlite
+// it is plain text).
+func (dbStore *DbStore) GetPasswordFunction() (string, error) {
+	switch dbStore.Config.Type {
+	case "mysql":
+		return "MD5(?)", nil
+	case "sqlite3":
+		return "?", nil
+	}
+	return "", errors.New(fmt.Sprintf("Unknown database: %s", dbStore.Config.Type))
+}
+
+func (dbStore *DbStore) DbStore() DbStore {
+	return *dbStore
+}
+
+func (dbStore *DbStore) getConnString() string {
+	info := dbStore.Config
+	switch info.Type {
+	case "sqlite3":
+		return info.Database
+	default:
+		return fmt.Sprintf("%s:%s@tcp(%s:%d)/%s", info.Username, info.Password, info.Host, info.Port, info.Database)
+	}
+}
+
+func (dbStore *DbStore) Connect() error {
+	if dbStore.Config == nil {
+		return errors.New("No configuration specified.")
+	}
+	connStr := dbStore.getConnString()
+	db, err := gorm.Open(dbStore.Config.Type, connStr)
+	if err != nil {
+		return err
+	}
+	dbStore.Db = &db
+	return nil
+}
+
+func (dbStore *DbStore) CreateSchema(force bool) error {
+	f := dbStore.createSchemaFuncs[dbStore.Config.Type]
+	if f == nil {
+		return errors.New(fmt.Sprintf("Unable to create schema for %s", dbStore.Config.Type))
+	}
+	return f(*dbStore, force)
+}
+
+// createSchemaMysql creates schema for a sqlite3 db
+func createSchemaSqlite3(dbStore DbStore, force bool) error {
+	var err error
+	schemaName := dbStore.Config.Database
+	if force {
+		_, err := os.Stat(schemaName)
+		if os.IsExist(err) {
+			err = os.Remove(schemaName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	connStr := dbStore.getConnString()
+	db, err := gorm.Open("sqlite3", connStr)
+
+	if err != nil {
+		return err
+	}
+
+	entities := dbStore.ServiceStore.Entities()
```

Comment:
Entity is very generic. ServiceStore doesn't tell me that this deals with schemas. So since all the names are totally generic here, when I see this line I have no idea what's going on. This line is not self-documenting.